# MovieLens Analytics

In this project, I'll analyze 45,000 movies from MovieLens Dataset consisting of movies up to July 2017 with the use of PostgreSQL and Pandas

data source: https://www.kaggle.com/datasets/rounakbanik/the-movies-dataset?select=movies_metadata.csv

To keep things more compact and readable, let's explore data and figure out what columns we'll use in later SQL queries with the help of Pandas

## Project structure:

This project will consist of the following parts (feel free to scroll to whatever part you wish to see first)

1. Data Preparation with Pandas

2. Loading Data to PostgreSQL

3. Establishing connection with the database in Jupyter Notebook

4. Writing SQL Queries

5. Summary

# Data Preparation with Pandas

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("data/movies_metadata.csv")
# Transpose for easier exploration of this dataset with many cols
df.head(3).transpose()

C:\Windows\Temp\ipykernel_1196\2753781476.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("data/movies_metadata.csv")


0  \
adult                                                              False   
belongs_to_collection  {'id': 10194, 'name': 'Toy Story Collection', ...   
budget                                                          30000000   
genres                 [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
homepage                            http://toystory.disney.com/toy-story   
id                                                                   862   
imdb_id                                                        tt0114709   
original_language                                                     en   
original_title                                                 Toy Story   
overview               Led by Woody, Andy's toys live happily in his ...   
popularity                                                     21.946943   
poster_path                             /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
production_companies      [{'name': 'Pixar Animation Studios', 'id': 3}]   
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...   
release_date                                                  1995-10-30   
revenue                                                      373554033.0   
runtime                                                             81.0   
spoken_languages                [{'iso_639_1': 'en', 'name': 'English'}]   
status                                                          Released   
tagline                                                              NaN   
title                                                          Toy Story   
video                                                              False   
vote_average                                                         7.7   
vote_count                                                        5415.0   

                                                                       1  \
adult                                                              False   
belongs_to_collection                                                NaN   
budget                                                          65000000   
genres                 [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
homepage                                                             NaN   
id                                                                  8844   
imdb_id                                                        tt0113497   
original_language                                                     en   
original_title                                                   Jumanji   
overview               When siblings Judy and Peter discover an encha...   
popularity                                                     17.015539   
poster_path                             /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
production_companies   [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...   
release_date                                                  1995-12-15   
revenue                                                      262797249.0   
runtime                                                            104.0   
spoken_languages       [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
status                                                          Released   
tagline                        Roll the dice and unleash the excitement!   
title                                                            Jumanji   
video                                                              False   
vote_average                                                         6.9   
vote_count                                                        2413.0   

                                                                       2  
adult                                                              False  
belongs_to_collection  {'id': 119050, 'name': 'Grumpy Old Men Collect...  
budget                                      

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [4]:
df["adult"].value_counts()

adult
False                                                                                                                             45454
True                                                                                                                                  9
 - Written by Ørnås                                                                                                                   1
 Rune Balot goes to a casino connected to the October corporation to try to wrap up her case once and for all.                        1
 Avalanche Sharks tells the story of a bikini contest that turns into a horrifying affair when it is hit by a shark avalanche.        1
Name: count, dtype: int64

In [5]:
df["video"].value_counts()

video
False    45367
True        93
Name: count, dtype: int64

In [6]:
df["status"].value_counts()

status
Released           45014
Rumored              230
Post Production       98
In Production         20
Planned               15
Canceled               2
Name: count, dtype: int64

The columns 'adult', 'status' and 'video' have predominantly one value, so let's remove them. Also, let's remove 'poster_path', 'hopepage' (not many not null values), 'spoken_languages', 'overview' and 'tagline' (we won't be conducting text analysis here + rows can get inconsistent with big amount of text from these cols)

In [7]:
df = df.drop(
    [
        "adult",
        "status",
        "video",
        "poster_path",
        "original_title",
        "homepage",
        "imdb_id",
        "spoken_languages",
        "overview",
        "tagline",
    ],
    axis=1,
)
df.head(3).transpose()

0  \
belongs_to_collection  {'id': 10194, 'name': 'Toy Story Collection', ...   
budget                                                          30000000   
genres                 [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
id                                                                   862   
original_language                                                     en   
popularity                                                     21.946943   
production_companies      [{'name': 'Pixar Animation Studios', 'id': 3}]   
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...   
release_date                                                  1995-10-30   
revenue                                                      373554033.0   
runtime                                                             81.0   
title                                                          Toy Story   
vote_average                                                         7.7   
vote_count                                                        5415.0   

                                                                       1  \
belongs_to_collection                                                NaN   
budget                                                          65000000   
genres                 [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
id                                                                  8844   
original_language                                                     en   
popularity                                                     17.015539   
production_companies   [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...   
release_date                                                  1995-12-15   
revenue                                                      262797249.0   
runtime                                                            104.0   
title                                                            Jumanji   
vote_average                                                         6.9   
vote_count                                                        2413.0   

                                                                       2  
belongs_to_collection  {'id': 119050, 'name': 'Grumpy Old Men Collect...  
budget                                                                 0  
genres                 [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...  
id                                                                 15602  
original_language                                                     en  
popularity                                                       11.7129  
production_companies   [{'name': 'Warner Bros.', 'id': 6194}, {'name'...  
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...  
release_date                                                  1995-12-22  
revenue                                                              0.0  
runtime                                                            101.0  
title                                                   Grumpier Old Men  
vote_average                                                         6.5  
vote_count                                                          92.0

##### IMPORTANT: strange cases like budget of 0 for 'Grumpier Old Men' at the right will be handled in later SQL queries

Now let's have a look at dtypes

## Converting dtypes to more appropriate ones

In [8]:
df.dtypes

belongs_to_collection     object
budget                    object
genres                    object
id                        object
original_language         object
popularity                object
production_companies      object
production_countries      object
release_date              object
revenue                  float64
runtime                  float64
title                     object
vote_average             float64
vote_count               float64
dtype: object

First of all, let's handle 'release_date' column

In [9]:
# Convert 'release_date' column to datetime type
df["release_date"] = pd.to_datetime(df["release_date"], errors="coerce")
# Format the 'release_date' column to display only the date part
df["release_date"] = df["release_date"].dt.strftime("%Y-%m-%d")
# Count the number of rows with bad date values
bad_date_count = df["release_date"].isnull().sum()
print(f"Number of rows with bad date values: {bad_date_count}")

Number of rows with bad date values: 90


Since 90 rows compared to 45,000 in total is nothing, we can freely remove them

In [10]:
# Remove rows with null or NaT values
df = df.dropna(subset=["release_date"])
bad_date_count = df["release_date"].isnull().sum()
print(f"Number of rows with bad date values: {bad_date_count}")

Number of rows with bad date values: 0


The column 'budget' contains non-numerical values like '/ff9qCepilowshEtG2GYWwzt2bs4.jpg'. Let's remove them

In [11]:
# Clean 'budget' column to remove non-numeric characters
df["budget"] = df["budget"].str.replace(r"\D", "", regex=True)

I don't like that columns with whole numbers like 'runtime' or 'vote_count' have dtype set to float. Let's change that

In [12]:
# Specify columns and their new data types
dict_columns_to_convert = {
    "budget": "int64",
    "revenue": "int64",
    "runtime": "int",
    "vote_count": "int",
    "id": "int"
}
# Clean 'budget' column to remove non-numeric characters
df["budget"] = df["budget"].str.replace(r"\D", "", regex=True)
# Fill NaN values with 0
cols_to_fill = list(dict_columns_to_convert.keys())
df[cols_to_fill] = df[cols_to_fill].fillna(0)
# Convert columns to integer type
df = df.astype(dict_columns_to_convert)
# Check the data types of the DataFrame
print(df.dtypes)

belongs_to_collection     object
budget                     int64
genres                    object
id                         int32
original_language         object
popularity                object
production_companies      object
production_countries      object
release_date              object
revenue                    int64
runtime                    int32
title                     object
vote_average             float64
vote_count                 int32
dtype: object


Time to extract data from JSON columns

## Working with 'belongs_to_collection' column

In [13]:
from ast import literal_eval

def extract_franchise_name(x):
    try:
        # Use literal_eval to safely evaluate the string as a Python dictionary
        # Extract the 'name' value from the dictionary
        return literal_eval(x)["name"]
    except (ValueError, TypeError):
        return None

# Apply the extract_franchise_name function to each value in the 'belongs_to_collection' column
df["franchise"] = df["belongs_to_collection"].apply(extract_franchise_name).str.strip()
# Remove the word 'Collection' (case-insensitive) from the end of each franchise name
df["franchise"] = df["franchise"].str.replace(r"[Cc]ollection$", "", regex=True)
# Remove trailing spaces before and after the string
df["franchise"] = df["franchise"].str.strip()
df = df.drop(["belongs_to_collection"], axis=1)

In [14]:
df["franchise"].value_counts().head(7)

franchise
The Bowery Boys                  29
Totò                             27
Zatôichi: The Blind Swordsman    26
James Bond                       26
The Carry On                     25
Pokémon                          22
Charlie Chan (Sidney Toler)      21
Name: count, dtype: int64

## Working with 'production_countries' column

In [15]:
def process_countries(countries):
    try:
        countries_list = literal_eval(countries)
        if len(countries_list) == 1:
            return countries_list[0]["name"]
        # Put 'Multiple' if movie was made in several countries
        elif len(countries_list) > 1:
            return "Multiple"
        # Put 'NULL' if the field is empty
        else:
            return None
    except (ValueError, TypeError):
        return None

# Apply the process_countries function to each value in the 'production_countries' column
df["production_country"] = df["production_countries"].apply(process_countries)
df = df.drop(["production_countries"], axis=1)

In [16]:
df["production_country"].value_counts().head(7)

production_country
United States of America    17846
Multiple                     7026
United Kingdom               2235
France                       1653
Japan                        1356
Italy                        1029
Canada                        840
Name: count, dtype: int64

## Working with 'genres' column

In [17]:
# Convert the stringified JSON into a list of dictionaries
df["genres"] = df["genres"].apply(
    lambda x: literal_eval(x.replace("'", '"')) if isinstance(x, str) else []
)
# Extract the names of genres into a list
df["genres"] = df["genres"].apply(
    lambda x: [genre["name"] for genre in x] if isinstance(x, list) else []
)
# Display the DataFrame with the extracted genre names
df[["title", "genres"]].head(3)

title                        genres
0         Toy Story   [Animation, Comedy, Family]
1           Jumanji  [Adventure, Fantasy, Family]
2  Grumpier Old Men             [Romance, Comedy]

In [18]:
# Flatten the list of genre names
flat_genre_names = [genre for sublist in df["genres"] for genre in sublist]
# Get the unique genre names
unique_genre_names = set(flat_genre_names)
# Print the unique genre names
print(f"There are {len(unique_genre_names)} unique genres.")
print(unique_genre_names)

There are 20 unique genres.
{'Action', 'Thriller', 'Comedy', 'Western', 'Documentary', 'Music', 'Romance', 'Science Fiction', 'War', 'Fantasy', 'History', 'TV Movie', 'Adventure', 'Animation', 'Foreign', 'Family', 'Horror', 'Drama', 'Mystery', 'Crime'}


We can see that 'genres' colomn has faulty data like 'Carousel Productions' or 'Vision View Entertainment', which sound like production companies, not genres. Thus, let's remove such values from the column

In [19]:
# Define the list of valid genre names
valid_genres = {
    "Action", "Adventure", "Animation", "Comedy", "Crime", "Documentary",
    "Drama", "Family", "Fantasy", "History", "Horror", "Music", "Mystery",
    "Romance", "Science Fiction", "Thriller", "War", "Western"
}
# Filter the genre_names column to include only the valid genres
df["genres"] = df["genres"].apply(lambda x: [genre for genre in x if genre in valid_genres])

Now let's check again

In [20]:
flat_genre_names = [genre for sublist in df["genres"] for genre in sublist]
unique_genre_names = set(flat_genre_names)
print(f"There are {len(unique_genre_names)} unique genres.")
print(unique_genre_names)

There are 18 unique genres.
{'Action', 'Western', 'Family', 'Horror', 'War', 'Thriller', 'Mystery', 'Romance', 'Music', 'Fantasy', 'History', 'Documentary', 'Comedy', 'Drama', 'Adventure', 'Animation', 'Science Fiction', 'Crime'}


Let's treat the first genre value as the main genre of the movie

In [21]:
# Create a separate column for the main genre
df["main_genre"] = df["genres"].apply(lambda x: x[0] if x else None)

One movie can belong to many genres and one genre can be applied to many movies. It's a many-to-many relationship. Ideally, this kind of relationship is supposed to be broken into two 1:M relationships and connected with an intermidiate or junction table. However, because

- it's a project to show my knowledge mainly of writing SQL queries
- I'm applying to a junior data analyst position, and, at that role, you're not supposed to design databases
- preparation part is already too long
- maximum string length for genres is known (80 symbols for the movie with the title 'Yu-Gi-Oh')

I'll keep things simple and connect genre names by comma.

To conclude our work with this column, let's sort values in 'genres' column alphabetically and then connect the values by a comma

In [22]:
# Extract the names of genres into a list and sort them alphabetically
df["genres"] = df["genres"].apply(
    lambda x: sorted([genre for genre in x]) if isinstance(x, list) else []
)
# Convert the list of genres into a string with comma as a delimiter
df["genres"] = df["genres"].apply(lambda x: ", ".join(x) if x else None)
df[["title", "genres", "main_genre"]].head(3)

title                      genres main_genre
0         Toy Story   Animation, Comedy, Family  Animation
1           Jumanji  Adventure, Family, Fantasy  Adventure
2  Grumpier Old Men             Comedy, Romance    Romance

In [23]:
df["genres"].value_counts().head(7)

genres
Drama                     5408
Comedy                    3729
Documentary               2861
Drama, Romance            1820
Comedy, Drama             1765
Comedy, Romance           1186
Comedy, Drama, Romance    1107
Name: count, dtype: int64

And so the time has come to work with the last JSON column

## Working with 'production_companies' column

In [24]:
def process_companies(companies):
    try:
        companies_list = literal_eval(companies)
        if len(companies_list) == 1:
            return companies_list[0]["name"]
        elif len(companies_list) > 1:
            return ", ".join([company["name"] for company in companies_list])
        else:
            return None
    except (ValueError, TypeError):
        return None

# Apply the process_companies function to each value in the 'production_companies' column
df["production_companies"] = df["production_companies"].apply(process_companies)
# Extract the text before the first comma as the main production company
df["main_production_company"] = df["production_companies"].apply(lambda x: x.split(",")[0] if isinstance(x, str) else None)
df[["title", "main_production_company", "production_companies"]].head()

title                 main_production_company  \
0                    Toy Story                 Pixar Animation Studios   
1                      Jumanji                        TriStar Pictures   
2             Grumpier Old Men                            Warner Bros.   
3            Waiting to Exhale  Twentieth Century Fox Film Corporation   
4  Father of the Bride Part II                   Sandollar Productions   

                                production_companies  
0                            Pixar Animation Studios  
1  TriStar Pictures, Teitler Film, Interscope Com...  
2                       Warner Bros., Lancaster Gate  
3             Twentieth Century Fox Film Corporation  
4         Sandollar Productions, Touchstone Pictures

In [25]:
df["main_production_company"].value_counts().head(10)

main_production_company
Paramount Pictures                        998
Metro-Goldwyn-Mayer (MGM)                 852
Twentieth Century Fox Film Corporation    780
Warner Bros.                              757
Universal Pictures                        754
Columbia Pictures                         429
Columbia Pictures Corporation             403
RKO Radio Pictures                        301
United Artists                            272
Walt Disney Pictures                      262
Name: count, dtype: int64

After all the operations completed above, we have the following df

In [26]:
df.head(3).transpose()

0  \
budget                                    30000000   
genres                   Animation, Comedy, Family   
id                                             862   
original_language                               en   
popularity                               21.946943   
production_companies       Pixar Animation Studios   
release_date                            1995-10-30   
revenue                                  373554033   
runtime                                         81   
title                                    Toy Story   
vote_average                                   7.7   
vote_count                                    5415   
franchise                                Toy Story   
production_country        United States of America   
main_genre                               Animation   
main_production_company    Pixar Animation Studios   

                                                                         1  \
budget                                                            65000000   
genres                                          Adventure, Family, Fantasy   
id                                                                    8844   
original_language                                                       en   
popularity                                                       17.015539   
production_companies     TriStar Pictures, Teitler Film, Interscope Com...   
release_date                                                    1995-12-15   
revenue                                                          262797249   
runtime                                                                104   
title                                                              Jumanji   
vote_average                                                           6.9   
vote_count                                                            2413   
franchise                                                             None   
production_country                                United States of America   
main_genre                                                       Adventure   
main_production_company                                   TriStar Pictures   

                                                    2  
budget                                              0  
genres                                Comedy, Romance  
id                                              15602  
original_language                                  en  
popularity                                    11.7129  
production_companies     Warner Bros., Lancaster Gate  
release_date                               1995-12-22  
revenue                                             0  
runtime                                           101  
title                                Grumpier Old Men  
vote_average                                      6.5  
vote_count                                         92  
franchise                              Grumpy Old Men  
production_country           United States of America  
main_genre                                    Romance  
main_production_company                  Warner Bros.

## Adding data from other two datasets

In [27]:
credits = pd.read_csv('data/credits.csv')
keywords = pd.read_csv('data/keywords.csv')
df = df.merge(credits, on='id')
df = df.merge(keywords, on='id')
df.head(3)

budget                      genres     id original_language popularity  \
0  30000000   Animation, Comedy, Family    862                en  21.946943   
1  65000000  Adventure, Family, Fantasy   8844                en  17.015539   
2         0             Comedy, Romance  15602                en    11.7129   

                                production_companies release_date    revenue  \
0                            Pixar Animation Studios   1995-10-30  373554033   
1  TriStar Pictures, Teitler Film, Interscope Com...   1995-12-15  262797249   
2                       Warner Bros., Lancaster Gate   1995-12-22          0   

   runtime             title  vote_average  vote_count       franchise  \
0       81         Toy Story           7.7        5415       Toy Story   
1      104           Jumanji           6.9        2413            None   
2      101  Grumpier Old Men           6.5          92  Grumpy Old Men   

         production_country main_genre  main_production_company  \
0  United States of America  Animation  Pixar Animation Studios   
1  United States of America  Adventure         TriStar Pictures   
2  United States of America    Romance             Warner Bros.   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...

## Extracting director

In [28]:
def extract_director(crew_list):
    for crew_member in crew_list:
        if crew_member["job"] == "Director":
            return crew_member["name"]
    return None

# For function get_director to work, convert
# the string representations to actual dictionaries
df["crew"] = df["crew"].apply(literal_eval)
# Extract the director's name for each movie
df["director"] = df["crew"].apply(extract_director)

In [29]:
df["director"].value_counts().head()

director
John Ford           67
Michael Curtiz      66
Werner Herzog       55
Julien Duvivier     54
Alfred Hitchcock    53
Name: count, dtype: int64

In [30]:
df[df["director"] == "Martin Scorsese"].head()

budget          genres     id original_language popularity  \
15    52000000    Crime, Drama    524                en  10.137389   
109    1300000    Crime, Drama    103                en  14.092713   
407   34000000  Drama, Romance  10436                en   8.013617   
1177  25000000    Crime, Drama    769                en  15.424092   
1192  18000000           Drama   1578                en   8.868558   

                                   production_companies release_date  \
15    Universal Pictures, Légende Entreprises, Syali...   1995-11-22   
109   Columbia Pictures Corporation, Italo/Judeo Pro...   1976-02-07   
407                                   Columbia Pictures   1993-09-17   
1177                                      Winkler Films   1990-09-12   
1192                                     United Artists   1980-11-14   

        revenue  runtime                 title  vote_average  vote_count  \
15    116112375      178                Casino           7.8        1343   
109    28262574      114           Taxi Driver           8.1        2632   
407    32255440      139  The Age of Innocence           7.0         172   
1177   46836394      145            GoodFellas           8.2        3211   
1192   23000000      129           Raging Bull           7.7         968   

     franchise        production_country main_genre  \
15        None                  Multiple      Drama   
109       None  United States of America      Crime   
407       None  United States of America      Drama   
1177      None  United States of America      Drama   
1192      None  United States of America      Drama   

            main_production_company  \
15               Universal Pictures   
109   Columbia Pictures Corporation   
407               Columbia Pictures   
1177                  Winkler Films   
1192                 United Artists   

                                                   cast  \
15    [{'cast_id': 4, 'character': "Sam 'Ace' Rothst...   
109   [{'cast_id': 5, 'character': 'Travis Bickle', ...   
407   [{'cast_id': 1, 'character': 'Newland Archer',...   
1177  [{'cast_id': 16, 'character': 'James Conway', ...   
1192  [{'cast_id': 1, 'character': 'Jake La Motta', ...   

                                                   crew  \
15    [{'credit_id': '52fe424dc3a36847f80139cd', 'de...   
109   [{'credit_id': '52fe4218c3a36847f80037c1', 'de...   
407   [{'credit_id': '52fe43709251416c750106b3', 'de...   
1177  [{'credit_id': '52fe4274c3a36847f801fcc3', 'de...   
1192  [{'credit_id': '52fe4300c3a36847f8032f25', 'de...   

                                               keywords         director  
15    [{'id': 383, 'name': 'poker'}, {'id': 726, 'na...  Martin Scorsese  
109   [{'id': 422, 'name': 'vietnam veteran'}, {'id'...  Martin Scorsese  
407   [{'id': 213, 'name': 'upper class'}, {'id': 24...  Martin Scorsese  
1177  [{'id': 378, 'name': 'prison'}, {'id': 818, 'n...  Martin Scorsese  
1192  [{'id': 396, 'name': 'transporter'}, {'id': 93...  Martin Scorsese

## Extracting top actors

In [31]:
def extract_top_actors(cast_list, top_n=3):
    top_actors = [actor["name"] for actor in cast_list[:top_n]]  # Select the top N actors
    return ", ".join(top_actors)

# Convert the string representations to actual dictionaries
df["cast"] = df["cast"].apply(literal_eval)
# Extract the top 3 actor names for each movie
df["top_3_actors"] = df["cast"].apply(lambda x: extract_top_actors(x, top_n=3))
# Extract the top 1 actor name for each movie
df["main_actor"] = df["cast"].apply(lambda x: extract_top_actors(x, top_n=1))

In [32]:
df[df["title"] == "The Empire Strikes Back"][
    ["title", "release_date", "director", "top_3_actors", "main_actor"]
]

title release_date        director  \
1161  The Empire Strikes Back   1980-05-17  Irvin Kershner   

                                   top_3_actors   main_actor  
1161  Mark Hamill, Harrison Ford, Carrie Fisher  Mark Hamill

## Extracting keywords

In [33]:
from collections import Counter
import pandas as pd

# Convert the string representations to actual dictionaries
df["keywords"] = df["keywords"].apply(literal_eval)
# Flatten the list of dictionaries in the 'keywords' column
keywords = [keyword["name"] for sublist in df["keywords"] for keyword in sublist]
# Count the frequencies of each keyword
keyword_counts = Counter(keywords)
# Sort the keywords based on their frequencies in descending order
sorted_keywords = sorted(keyword_counts.items(), key=lambda x: x[1], reverse=True)
# Remove keywords that rarely occur
sorted_keywords_filtered = [(keyword, count) for keyword, count in sorted_keywords if count > 9]
# Create a set of keywords that appear in sorted_keywords_filtered
filtered_keywords_set = set([keyword for keyword, _ in sorted_keywords_filtered])
# Print the sorted and filtered keywords
for keyword, count in sorted_keywords_filtered[:10]:
    print(f"{keyword}: {count}")

woman director: 3125
independent film: 1942
murder: 1313
based on novel: 837
musical: 733
sex: 685
violence: 651
biography: 641
nudity: 636
revenge: 626


In [34]:
# Print the number of unique keywords
num_unique_keywords_before = len(keyword_counts)
print(f"Number of unique keywords before filtering: {num_unique_keywords_before}")
num_unique_keywords_after = len(sorted_keywords_filtered)
print(f"Number of unique keywords after filtering: {num_unique_keywords_after}")

Number of unique keywords before filtering: 19954
Number of unique keywords after filtering: 2957


In [35]:
def filter_keywords(keywords_list):
    filtered_keywords = []
    for keyword in keywords_list:
        if keyword["name"] in filtered_keywords_set and len(filtered_keywords) < 5:
            filtered_keywords.append(keyword["name"])
            if len(filtered_keywords) == 5:
                break
    return ", ".join(filtered_keywords)

# Add a new column 'filtered_keywords' to the DataFrame
df["filtered_keywords"] = df["keywords"].apply(filter_keywords)

In [36]:
df[["keywords", "filtered_keywords"]]

keywords  \
0      [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1      [{'id': 10090, 'name': 'board game'}, {'id': 1...   
2      [{'id': 1495, 'name': 'fishing'}, {'id': 12392...   
3      [{'id': 818, 'name': 'based on novel'}, {'id':...   
4      [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...   
...                                                  ...   
46535                                                 []   
46536  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...   
46537                                                 []   
46538                                                 []   
46539                                                 []   

                                       filtered_keywords  
0                jealousy, toy, boy, friendship, friends  
1      disappearance, based on children's book, new h...  
2             fishing, best friend, duringcreditsstinger  
3      based on novel, interracial relationship, sing...  
4      baby, midlife crisis, confidence, aging, daughter  
...                                                  ...  
46535                                                     
46536                                artist, play, pinoy  
46537                                                     
46538                                                     
46539                                                     

[46540 rows x 2 columns]

## Final steps of data preparation with Pandas

As a final touch, let's redo 'id' column, drop several not needed anymore columns, change dtypes, and rearrange columns

In [37]:
# Drop columns 'id', 'cast', 'crew', and 'keywords'
df = df.drop(columns=["id", "cast", "crew", "keywords"])
# Rename 'filtered_keywosrds' column to 'keywords'
df = df.rename(columns={"filtered_keywords": "keywords"})
# Drop dublicate rows
df = df.drop_duplicates()
# Add the new 'id' column as the first column
df.insert(0, "id", range(1, 1 + len(df)))
df.head().transpose()
new_cols_order = [
    "id",
    "title",
    "franchise",
    "release_date",
    "runtime",
    "main_genre",
    "genres",
    "production_country",
    "main_production_company",
    "production_companies",
    "original_language",
    "budget",
    "revenue",
    "popularity",
    "vote_average",
    "vote_count",
    "director",
    "main_actor",
    "top_3_actors",
    "keywords"
]
df = df[new_cols_order]
df.head(3).transpose()

0  \
id                                                             1   
title                                                  Toy Story   
franchise                                              Toy Story   
release_date                                          1995-10-30   
runtime                                                       81   
main_genre                                             Animation   
genres                                 Animation, Comedy, Family   
production_country                      United States of America   
main_production_company                  Pixar Animation Studios   
production_companies                     Pixar Animation Studios   
original_language                                             en   
budget                                                  30000000   
revenue                                                373554033   
popularity                                             21.946943   
vote_average                                                 7.7   
vote_count                                                  5415   
director                                           John Lasseter   
main_actor                                             Tom Hanks   
top_3_actors                   Tom Hanks, Tim Allen, Don Rickles   
keywords                 jealousy, toy, boy, friendship, friends   

                                                                         1  \
id                                                                       2   
title                                                              Jumanji   
franchise                                                             None   
release_date                                                    1995-12-15   
runtime                                                                104   
main_genre                                                       Adventure   
genres                                          Adventure, Family, Fantasy   
production_country                                United States of America   
main_production_company                                   TriStar Pictures   
production_companies     TriStar Pictures, Teitler Film, Interscope Com...   
original_language                                                       en   
budget                                                            65000000   
revenue                                                          262797249   
popularity                                                       17.015539   
vote_average                                                           6.9   
vote_count                                                            2413   
director                                                      Joe Johnston   
main_actor                                                  Robin Williams   
top_3_actors                  Robin Williams, Jonathan Hyde, Kirsten Dunst   
keywords                 disappearance, based on children's book, new h...   

                                                                  2  
id                                                                3  
title                                              Grumpier Old Men  
franchise                                            Grumpy Old Men  
release_date                                             1995-12-22  
runtime                                                         101  
main_genre                                                  Romance  
genres                                              Comedy, Romance  
production_country                         United States of America  
main_production_company                                Warner Bros.  
production_companies                   Warner Bros., Lancaster Gate  
original_language                                                en  
budget                                                            0  
revenue                                                           0  
popularity     

Columns explanation:
- id - row id
- title - official title of the movie
- franchise - a particular franchise to which the movie belongs, if applicable
- release_year - theatrical release year of the movie
- runtime - movie duration/runtime in minutes
- main_genre - the first/main genre of the movie
- genres - genres associated with the movie, separated by a comma
- production_country - the country/countries where the movie was shot/produced. If several countries were involved, the cell contains the value 'Multiple'
- main_production_company - main production company
- production_companies - production companies involved in making of the movie
- original_language - the language in which the movie was originally shot
- budget - movie budget in dollars
- revenue - total movie revenue in dollars
- popularity - popularity score assigned by TMDB
- vote_average - average movie rating
- vote_count - number of votes by users, counted by TMDB
- director - name of the movie director
- main_actor - name of the actor that played the main role
- top_3_actors - names of top 3 actors in the movie
- keywords - keywords associated with the movie

Let's save the cleaned up dataset, which we'll use in the next chapters

In [38]:
# df.to_csv("data/data_expanded.csv", index=False)

# Loading Data to PostgreSQL

Okay, time to finally work with the database

## Creating a new database

<img src="images/creating_database_1.png">

<img src="images/creating_database_2.png">

<img src="images/creating_database_3.png">

Database is successfully created. Now time to create a table for our movies data

## Creating 'movies' table

<img src="images/creating_table_1.png">

And to the last step - filling the newly created table with data

## Filling table with data

### Solution 1 - inside pgAdmin4

<img src="images/filling_table_1.png">

<img src="images/filling_table_2.png">

<img src="images/filling_table_3.png">

<img src="images/filling_table_4.png">

<img src="images/filling_table_5.png">

### Solution 2 - using to_sql function from Pandas library

In [42]:
import psycopg2
from sqlalchemy import create_engine
import configparser
import pandas as pd

config = configparser.ConfigParser()
config.read("db_vars.cfg")
DB_USER = config.get("SQL", "DB_USER")
DB_PASSWORD = config.get("SQL", "DB_PASSWORD")
DB_NAME = "movielens"
conn_string = f"postgresql://{DB_USER}:{DB_PASSWORD}@localhost/{DB_NAME}"
engine = create_engine(conn_string)

df.to_sql("movies", engine, if_exists="append", index=False)

376

# Establishing connection with the database in Jupyter Notebook

To be able to connect to our database, we'll need the following libraries:

1. **ipython-sql**: Enables executing SQL queries in Jupyter Notebooks for direct database interaction.

2. **psycopg2**: Python adapter for PostgreSQL, facilitating PostgreSQL database interaction.

3. **configparser**: Manages configuration settings, like database credentials, in a more secure way.

In [1]:
import psycopg2
import configparser
import pandas as pd
# magic function used to connect with the postgresql database
%load_ext sql 

Read the parameters from the config file

In [2]:
config = configparser.ConfigParser()
config.read("db_vars.cfg")
DB_USER = config.get("SQL", "DB_USER")
DB_PASSWORD = config.get("SQL", "DB_PASSWORD")

Connect to the database

In [3]:
DB_NAME = "movielens"
conn_string = f"postgresql://{DB_USER}:{DB_PASSWORD}@127.0.0.1/{DB_NAME}"
%sql $conn_string

# Writing SQL Queries

After all these steps, we can finaly begin writing SQL queries

In [4]:
%%sql

SELECT *
FROM movies
LIMIT 5;

 * postgresql://postgres:***@127.0.0.1/movielens
5 rows affected.


[(1, 'Toy Story', 'Toy Story', datetime.date(1995, 10, 30), 81, 'Animation, Comedy, Family', 'United States of America', 'Pixar Animation Studios', 'en', 30000000, 373554033, 21.946943, 7.7, 5415),
 (2, 'Jumanji', None, datetime.date(1995, 12, 15), 104, 'Adventure, Family, Fantasy', 'United States of America', 'TriStar Pictures, Teitler Film, Interscope Communications', 'en', 65000000, 262797249, 17.015539, 6.9, 2413),
 (3, 'Grumpier Old Men', 'Grumpy Old Men', datetime.date(1995, 12, 22), 101, 'Comedy, Romance', 'United States of America', 'Warner Bros., Lancaster Gate', 'en', 0, 0, 11.7129, 6.5, 92),
 (4, 'Waiting to Exhale', None, datetime.date(1995, 12, 22), 127, 'Comedy, Drama, Romance', 'United States of America', 'Twentieth Century Fox Film Corporation', 'en', 16000000, 81452156, 3.859495, 6.1, 34),
 (5, 'Father of the Bride Part II', 'Father of the Bride', datetime.date(1995, 2, 10), 106, 'Comedy', 'United States of America', 'Sandollar Productions, Touchstone Pictures', 'en', 0, 76578911, 8.387519, 5.7, 173)]

There are still some imperfections. For example, values in revenue are too big to adequately distinguish or column names are sometimes too long. Thus, here arises a wonderful opportunity to demostrate the use of VIEW and the knowledge of ROUND and AS commands

## Creating VIEW


Hmm, the 'budget' and 'revenue' columns contain very large values, making it difficult to distinguish between them.. Thus, to improve readability, let's divide the values in the 'revenue' and 'budget' columns by one million, as movie budgets and revenues are typically counted in millions.

Also, let's round 'popularity' column to 2 decimal places for better readability

In [5]:
%%sql

SELECT
    title,
    release_date,
    budget,
    ROUND(budget / 1000000::NUMERIC, 2) AS budget_in_millions,
    revenue,
    ROUND(revenue / 1000000::NUMERIC, 2) AS revenue_in_millions,
    popularity,
    ROUND(popularity::NUMERIC, 2) AS popularity_rounded
FROM movies
LIMIT 5;

 * postgresql://postgres:***@127.0.0.1/movielens
5 rows affected.


[('Toy Story', datetime.date(1995, 10, 30), 30000000, Decimal('30.00'), 373554033, Decimal('373.55'), 21.946943, Decimal('21.95')),
 ('Jumanji', datetime.date(1995, 12, 15), 65000000, Decimal('65.00'), 262797249, Decimal('262.80'), 17.015539, Decimal('17.02')),
 ('Grumpier Old Men', datetime.date(1995, 12, 22), 0, Decimal('0.00'), 0, Decimal('0.00'), 11.7129, Decimal('11.71')),
 ('Waiting to Exhale', datetime.date(1995, 12, 22), 16000000, Decimal('16.00'), 81452156, Decimal('81.45'), 3.859495, Decimal('3.86')),
 ('Father of the Bride Part II', datetime.date(1995, 2, 10), 0, Decimal('0.00'), 76578911, Decimal('76.58'), 8.387519, Decimal('8.39'))]

Everything else appears to be fine except long column names and the column 'original_language', which we'll shorten to just 'language'. Time to create a VIEW called 'mv' (i.e., movies view) that will incorporate all the improvements we've made

In [6]:
%%sql

CREATE VIEW mv AS
SELECT
    id,
    title,
    franchise,
    release_date,
    runtime,
    genres,
    production_country AS prod_country,
    productioN_companies AS prod_companies,
    original_language AS lang,
    ROUND(budget / 1000000::NUMERIC, 2) AS budget_in_millions,
    ROUND(revenue / 1000000::NUMERIC, 2) AS revenue_in_millions,
    ROUND(popularity::NUMERIC, 2) AS popularity,
    vote_average AS vote_avg,
    vote_count
FROM movies;

 * postgresql://postgres:***@127.0.0.1/movielens
(psycopg2.errors.DuplicateTable) relation "mv" already exists

[SQL: CREATE VIEW mv AS
SELECT
    id,
    title,
    franchise,
    release_date,
    runtime,
    genres,
    production_country AS prod_country,
    productioN_companies AS prod_companies,
    original_language AS lang,
    ROUND(budget / 1000000::NUMERIC, 2) AS budget_in_millions,
    ROUND(revenue / 1000000::NUMERIC, 2) AS revenue_in_millions,
    ROUND(popularity::NUMERIC, 2) AS popularity,
    vote_average AS vote_avg,
    vote_count
FROM movies;]
(Background on this error at: https://sqlalche.me/e/20/f405)


In [7]:
%%sql

-- DROP VIEW mv;

 * postgresql://postgres:***@127.0.0.1/movielens
(psycopg2.ProgrammingError) can't execute an empty query
[SQL: -- DROP VIEW mv;]
(Background on this error at: https://sqlalche.me/e/20/f405)


In [8]:
%%sql

SELECT *
FROM mv
LIMIT 5;

 * postgresql://postgres:***@127.0.0.1/movielens
5 rows affected.


[(1, 'Toy Story', 'Toy Story', datetime.date(1995, 10, 30), 81, 'Animation, Comedy, Family', 'United States of America', 'Pixar Animation Studios', 'en', Decimal('30.00'), Decimal('373.55'), Decimal('21.95'), 7.7, 5415),
 (2, 'Jumanji', None, datetime.date(1995, 12, 15), 104, 'Adventure, Family, Fantasy', 'United States of America', 'TriStar Pictures, Teitler Film, Interscope Communications', 'en', Decimal('65.00'), Decimal('262.80'), Decimal('17.02'), 6.9, 2413),
 (3, 'Grumpier Old Men', 'Grumpy Old Men', datetime.date(1995, 12, 22), 101, 'Comedy, Romance', 'United States of America', 'Warner Bros., Lancaster Gate', 'en', Decimal('0.00'), Decimal('0.00'), Decimal('11.71'), 6.5, 92),
 (4, 'Waiting to Exhale', None, datetime.date(1995, 12, 22), 127, 'Comedy, Drama, Romance', 'United States of America', 'Twentieth Century Fox Film Corporation', 'en', Decimal('16.00'), Decimal('81.45'), Decimal('3.86'), 6.1, 34),
 (5, 'Father of the Bride Part II', 'Father of the Bride', datetime.date(1995, 2, 10), 106, 'Comedy', 'United States of America', 'Sandollar Productions, Touchstone Pictures', 'en', Decimal('0.00'), Decimal('76.58'), Decimal('8.39'), 5.7, 173)]

As a reminder of what each column stands for, let's bring columns explanations with the incorporation of recent changes

Columns explanation:
- id - row id
- title - official title of the movie
- franchise - a particular franchise to which the movie belongs, if applicable
- release_date - theatrical release date of the movie
- runtime - movie duration/runtime in minutes
- genres - genres associated with the movie, separated by a comma
- prod_country - the country/countries where the movie was shot/produced. If several countries were involved, the cell contains the value 'Multiple'
- prod_companies - production companies involved in making of the movie
- lang - the language in which the movie was originally shot
- budget_in_millions - movie budget in millions of dollars
- revenue_in_millions - total movie revenue in millions of dollars
- popularity - popularity score assigned by TMDB
- vote_avg - average movie rating
- vote_count - number of votes by users, counted by TMDB

Now that the data is in satisfactory shape, we can finally start asking questions and trying to find answers to them.

## Question - Answer

### Q: Per each decade, find top 3 most popular movies from the middle of the 20th century (1940-69)

In [9]:
%%sql

WITH step1 AS (
    SELECT
        title,
        CASE EXTRACT(DECADE FROM release_date)
            WHEN 194 THEN '40s'
            WHEN 195 THEN '50s'
            WHEN 196 THEN '60s'
        END AS decade,
        popularity,
        ROW_NUMBER() OVER (
            PARTITION BY EXTRACT(DECADE FROM release_date)
            ORDER BY popularity DESC
        ) AS rn
    FROM mv
    WHERE EXTRACT(YEAR FROM release_date) BETWEEN 1940 AND 1969
)
SELECT
    title,
    decade,
    popularity
FROM step1
WHERE rn <= 3
ORDER BY decade, rn;

 * postgresql://postgres:***@127.0.0.1/movielens
9 rows affected.


[('Citizen Kane', '40s', Decimal('15.81')),
 ("It's a Wonderful Life", '40s', Decimal('15.03')),
 ('Dumbo', '40s', Decimal('14.66')),
 ('Vertigo', '50s', Decimal('18.21')),
 ('On the Waterfront', '50s', Decimal('18.21')),
 ('Rear Window', '50s', Decimal('17.91')),
 ('Psycho', '60s', Decimal('36.83')),
 ('The Jungle Book', '60s', Decimal('22.85')),
 ('2001: A Space Odyssey', '60s', Decimal('22.49'))]

Observation:

- For the 40s and 50s, the top 3 movies have very close popularity values, while the movie 'Psycho' from the 60s has a popularity value that is about 1.6 (37/23) times higher than the second and third most popular moveis of that decade.

### Q: Count how many movies were released in each month. Determine the month (or months) with the highest number of movie releases

In [10]:
%%sql

SELECT
    -- Represent month as a word instead of a number
    TO_CHAR(TO_DATE(EXTRACT(MONTH FROM release_date)::TEXT, 'MM'), 'Month') AS release_month,
    COUNT(id) AS movie_count,
    ROUND((COUNT(*)::NUMERIC / (SELECT COUNT(*) FROM mv) * 100), 2) AS movie_percentage
FROM mv
GROUP BY EXTRACT(MONTH FROM release_date)
ORDER BY EXTRACT(MONTH FROM release_date) ASC;

 * postgresql://postgres:***@127.0.0.1/movielens
12 rows affected.


[('January  ', 5912, Decimal('13.03')),
 ('February ', 3032, Decimal('6.68')),
 ('March    ', 3553, Decimal('7.83')),
 ('April    ', 3453, Decimal('7.61')),
 ('May      ', 3339, Decimal('7.36')),
 ('June     ', 3153, Decimal('6.95')),
 ('July     ', 2640, Decimal('5.82')),
 ('August   ', 3394, Decimal('7.48')),
 ('September', 4838, Decimal('10.66')),
 ('October  ', 4615, Decimal('10.17')),
 ('November ', 3661, Decimal('8.07')),
 ('December ', 3786, Decimal('8.34'))]

__A: January (13%), closely followed by September and October (11% and 10%), has the highest number of released movies.__

This finding coincides with the fact that in the film-making industry, these months are considered to be dump months:

"January, February, and the months following summer (especially September) have fewer high-profile movies because people are going to the movies less due to work and school, but the studios still need to release movies." (source: https://movieweb.com/january-movies-dump-month/)

Okay, now that we learned about dump months, let's try to find the answer to the following question

### Q: Find the months when high revenue movies (100+ million USD) are usualy released

In [11]:
%%sql

SELECT
    -- Represent month as a word instead of a number
    TO_CHAR(TO_DATE(EXTRACT(MONTH FROM release_date)::TEXT, 'MM'), 'Month') AS release_month,
    COUNT(id) AS movie_count,
    ROUND((COUNT(*)::NUMERIC / (SELECT COUNT(*) FROM mv WHERE revenue_in_millions > 100) * 100), 2) AS movie_percentage
FROM mv
WHERE revenue_in_millions > 100
GROUP BY EXTRACT(MONTH FROM release_date)
ORDER BY EXTRACT(MONTH FROM release_date) ASC;

 * postgresql://postgres:***@127.0.0.1/movielens
12 rows affected.


[('January  ', 44, Decimal('3.22')),
 ('February ', 79, Decimal('5.78')),
 ('March    ', 96, Decimal('7.02')),
 ('April    ', 70, Decimal('5.12')),
 ('May      ', 134, Decimal('9.80')),
 ('June     ', 185, Decimal('13.53')),
 ('July     ', 152, Decimal('11.12')),
 ('August   ', 88, Decimal('6.44')),
 ('September', 85, Decimal('6.22')),
 ('October  ', 96, Decimal('7.02')),
 ('November ', 137, Decimal('10.02')),
 ('December ', 201, Decimal('14.70'))]

__A: December (15%), June (14%), and July (11%) have the highest number of high revenue movies.__

This finding comes as no surprise because

"The most lucrative times of year for Hollywood are when schools are on break. As such, most major blockbusters and films expected to be box office hits are released during the Summer May-August window and the second half of December." (source: https://movieweb.com/january-movies-not-good/)

Since we're already on high revenue movies and dates, let's find out on what day of week they are usualy released

### Q: Find on what day of week high revenue movies get released

In [12]:
%%sql

SELECT
    TO_CHAR(release_date, 'Day') AS release_day_of_week,
    COUNT(id) AS movie_count,
    ROUND((COUNT(*)::NUMERIC / (SELECT COUNT(*) FROM mv WHERE revenue_in_millions > 100) * 100), 2) AS movie_percentage
FROM mv
WHERE revenue_in_millions > 100
GROUP BY EXTRACT(ISODOW FROM release_date), release_day_of_week
ORDER BY EXTRACT(ISODOW FROM release_date) ASC;

 * postgresql://postgres:***@127.0.0.1/movielens
7 rows affected.


[('Monday   ', 57, Decimal('4.17')),
 ('Tuesday  ', 122, Decimal('8.92')),
 ('Wednesday', 331, Decimal('24.21')),
 ('Thursday ', 348, Decimal('25.46')),
 ('Friday   ', 425, Decimal('31.09')),
 ('Saturday ', 48, Decimal('3.51')),
 ('Sunday   ', 36, Decimal('2.63'))]

__A: Friday (31%) with Thursday (25%) and Wednesday (24%) dominate here. In other words, staggering 80% of high revenue movies are released in the middle of the week (Wednesday - Friday).__

### Q: List Star Wars movies

And let's pretent that Disney didn't make any SW movies

In [13]:
%%sql

SELECT
    title, release_date, runtime, genres,
    prod_companies, budget_in_millions,
    revenue_in_millions, popularity, vote_avg, vote_count
FROM mv
WHERE franchise = 'Star Wars'
    -- let's pretend that Disney didn't make any SW movies
    AND EXTRACT(YEAR FROM release_date) < 2010;

 * postgresql://postgres:***@127.0.0.1/movielens
6 rows affected.


[('Star Wars', datetime.date(1977, 5, 25), 121, 'Action, Adventure, Science Fiction', 'Lucasfilm, Twentieth Century Fox Film Corporation', Decimal('11.00'), Decimal('775.40'), Decimal('42.15'), 8.1, 6778),
 ('The Empire Strikes Back', datetime.date(1980, 5, 17), 124, 'Action, Adventure, Science Fiction', 'Lucasfilm, Twentieth Century Fox Film Corporation', Decimal('18.00'), Decimal('538.40'), Decimal('19.47'), 8.2, 5998),
 ('Return of the Jedi', datetime.date(1983, 5, 23), 135, 'Action, Adventure, Science Fiction', 'Lucasfilm, Twentieth Century Fox Film Corporation', Decimal('32.35'), Decimal('572.70'), Decimal('14.59'), 7.9, 4763),
 ('Star Wars: Episode I - The Phantom Menace', datetime.date(1999, 5, 19), 136, 'Action, Adventure, Science Fiction', 'Lucasfilm', Decimal('115.00'), Decimal('924.32'), Decimal('15.65'), 6.4, 4526),
 ('Star Wars: Episode II - Attack of the Clones', datetime.date(2002, 5, 15), 142, 'Action, Adventure, Science Fiction', 'Lucasfilm', Decimal('120.00'), Decimal('649.40'), Decimal('14.07'), 6.4, 4074),
 ('Star Wars: Episode III - Revenge of the Sith', datetime.date(2005, 5, 17), 140, 'Action, Adventure, Science Fiction', 'Lucasfilm', Decimal('113.00'), Decimal('850.00'), Decimal('13.17'), 7.1, 4200)]

Observations:

- All Star Wars movies were released in the second half of May.

- The first trilogy was released with the help of 20th Century Fox, while the second one was made solely by Lucasfilm. Proposition: perhaps due to the huge success of the first trilogy, Lucasfilm could later afford to make movies on its own.

- In each trilogy, the second/middle movie brought in the least revenue

### Q: Find top 7 movie franchises by revenue

In [14]:
%%sql

SELECT
    franchise,
    COUNT(id) AS movie_count,
    SUM(revenue_in_millions) AS total_revenue_in_millions,
    ROUND(AVG(revenue_in_millions), 2) AS avg_revenue_per_movie_in_millions
FROM mv
WHERE franchise IS NOT NULL
GROUP BY franchise
ORDER BY total_revenue_in_millions DESC
LIMIT 7;

 * postgresql://postgres:***@127.0.0.1/movielens
7 rows affected.


[('Harry Potter', 8, Decimal('7707.37'), Decimal('963.42')),
 ('Star Wars', 8, Decimal('7434.50'), Decimal('929.31')),
 ('James Bond', 26, Decimal('7106.96'), Decimal('273.34')),
 ('The Fast and the Furious', 8, Decimal('5125.09'), Decimal('640.64')),
 ('Pirates of the Caribbean', 5, Decimal('4521.57'), Decimal('904.31')),
 ('Transformers', 5, Decimal('4366.11'), Decimal('873.22')),
 ('Despicable Me', 6, Decimal('3691.06'), Decimal('615.18'))]

Observations:

- Harry Potter franchise made the most amount of money (7.7 billion USD) and has the highest average revenue (963 million USD)

- The total revenue for the first three franchises is closely grouped together and (7.4 billion USD +- 0.3 billion USD). Then, between the top 3 and top 4 franchises, there is a big droup of 2 billion USD.

### Q: Find the number and percentage of movies with runtimes of up to an hour, two hours, three hours, and more

In [15]:
%%sql

WITH step1 AS (
    SELECT
        id,
        runtime,
        CASE
            WHEN runtime > 0 AND runtime < 61 THEN '1 hour'
            WHEN runtime > 60 AND runtime < 121 THEN '2 hours'
            WHEN runtime > 120 AND runtime < 181 THEN '3 hours'
            ELSE 'More'
        END AS runtime_group
    FROM mv
    WHERE runtime > 0
)
SELECT
    runtime_group,
    COUNT(id) AS count,
    ROUND((COUNT(id)::NUMERIC / (SELECT COUNT(*) FROM mv)) * 100, 2) AS percentage,
    ROUND(AVG(runtime), 2) AS avg_runtime_per_group
FROM step1
GROUP BY runtime_group
ORDER BY runtime_group ASC;


 * postgresql://postgres:***@127.0.0.1/movielens
4 rows affected.


[('1 hour', 2615, Decimal('5.76'), Decimal('31.57')),
 ('2 hours', 35975, Decimal('79.28'), Decimal('94.75')),
 ('3 hours', 4489, Decimal('9.89'), Decimal('137.89')),
 ('More', 516, Decimal('1.14'), Decimal('271.63'))]

Observations:

- Movies running from 61 to 120 minutes dominate this dataset, accounting for 79% of the total, with an average runtime of 95 minutes. This finding aligns with my (and probably of most people's) movie-watching experience, where the majority of films run for 90-120 minutes.

- The second place (10%) is taken by movies that run up to three hours. Interestingly, the average runtime here is 138 minutes, suggesting that many of these movies are prolonged two-hour-long films.

### Q: Select four major movie studios and provide a summary of their data

In [16]:
%%sql

WITH company_stats AS (
    SELECT
        CASE
            WHEN prod_companies LIKE '%Universal Pictures%' THEN 'Universal Pictures'
            WHEN prod_companies LIKE '%Century Fox%' THEN '20th Century Fox'
            WHEN prod_companies LIKE '%Pixar%' THEN 'Pixar'
            WHEN prod_companies LIKE '%Warner Bros%' THEN 'Warner Bros.'
        END AS movie_company,
        COUNT(*) AS movie_count,
        SUM(revenue_in_millions) AS total_revenue_in_millions,
        ROUND(AVG(revenue_in_millions), 2) AS avg_revenue_in_millions,
        MIN(EXTRACT(YEAR FROM release_date)) AS earliest_movie
    FROM mv
    WHERE revenue_in_millions > 0
    GROUP BY movie_company
)
SELECT *
FROM company_stats
WHERE movie_company IS NOT NULL
ORDER BY movie_company ASC;

 * postgresql://postgres:***@127.0.0.1/movielens
4 rows affected.


[('20th Century Fox', 358, Decimal('52546.19'), Decimal('146.78'), Decimal('1938')),
 ('Pixar', 18, Decimal('11188.53'), Decimal('621.59'), Decimal('1995')),
 ('Universal Pictures', 464, Decimal('55420.38'), Decimal('119.44'), Decimal('1931')),
 ('Warner Bros.', 509, Decimal('66469.21'), Decimal('130.59'), Decimal('1928'))]

Observations:

- Warner Bros. has the highest movie count and total revenue.

- Warner Bros. released their first movie earliest among the selected companies - in 1928.

- Pixar released their first movie latest - in 1995.

- Pixar produced the smallest number of movies (only 18 compared to hundreds from other companies).

- Pixar has the highest average revenue, 622 millions USD, which is more than 4 times higher (622/147) than the second highest of 20th Century Fox, which averages 147 million USD.

### Q: Find top 4 most spectacular movie successes and failures by profit

"The general rule of thumb is that movies need to make 2.5 times their budgets to turn a profit, as marketing costs are typically equal to production budgets, and then movie theaters get a cut of ticket sales."
(source: https://screenrant.com/fast-x-box-office-breakdown-success-flop/#:~:text=The%20general%20rule%20of%20thumb,a%20cut%20of%20ticket%20sales.)

However, because we're dealing mostly with blockbusters that cost 100+ millions of USD, we'll make it 1.5 instead of 2.5.


In [17]:
%%sql

WITH step1 AS (
    SELECT
        title,
        EXTRACT(YEAR FROM release_date) AS release_year,
        ROUND(1.5 * budget_in_millions, 2) AS appr_expenses_in_millions,
        revenue_in_millions,
        ROUND((revenue_in_millions - 1.5 * budget_in_millions), 2) AS profit_in_millions
    FROM mv
    -- remove empty or wrong values that can influece the result
    WHERE budget_in_millions >= 1 AND revenue_in_millions >= 1
), step2 AS (
    SELECT
        *,
        CASE
            WHEN profit_in_millions > 0 THEN 'Success'
            ELSE 'Failure'
        END AS profit_state,
        ROW_NUMBER() OVER (ORDER BY profit_in_millions DESC) AS rank_highest_profit,
        ROW_NUMBER() OVER (ORDER BY profit_in_millions ASC) AS rank_lowest_profit
    FROM step1
)
SELECT
    title,
    release_year,
    profit_state AS state,
    appr_expenses_in_millions,
    revenue_in_millions,
    profit_in_millions
FROM step2
WHERE rank_highest_profit <= 4 OR rank_lowest_profit <= 4
ORDER BY profit_in_millions ASC;

 * postgresql://postgres:***@127.0.0.1/movielens
8 rows affected.


[('The Lone Ranger', Decimal('2013'), 'Failure', Decimal('382.50'), Decimal('89.29'), Decimal('-293.21')),
 ('Valerian and the City of a Thousand Planets', Decimal('2017'), 'Failure', Decimal('296.21'), Decimal('90.02'), Decimal('-206.19')),
 ('The Alamo', Decimal('2004'), 'Failure', Decimal('217.50'), Decimal('25.82'), Decimal('-191.68')),
 ('Mars Needs Moms', Decimal('2011'), 'Failure', Decimal('225.00'), Decimal('38.99'), Decimal('-186.01')),
 ('Jurassic World', Decimal('2015'), 'Success', Decimal('225.00'), Decimal('1513.53'), Decimal('1288.53')),
 ('Titanic', Decimal('1997'), 'Success', Decimal('300.00'), Decimal('1845.03'), Decimal('1545.03')),
 ('Star Wars: The Force Awakens', Decimal('2015'), 'Success', Decimal('367.50'), Decimal('2068.22'), Decimal('1700.72')),
 ('Avatar', Decimal('2009'), 'Success', Decimal('355.50'), Decimal('2787.97'), Decimal('2432.47'))]

Observations:

- 7/8 movies in this list were released in the 21st century.

- _The Lone Ranger_ from 2013 lost the most amount of money - 293 million USD.

- _Avatar_ from 2009 made the highest profit - 2.4 billion USD.

# Summary

In this project, I

1. Prepared data with Pandas

2. Created a PostgreSQL database and table with pgAdmin4

3. Filled the table with data

4. Connected to the database in Jupyter Notebook

5. Wrote about ten SQL queries of varying difficulty and provided comments for the results